matchms/SpectrumProcessor.py

```diff
@@ -0,0 +1,287 @@
+        """
+        Add a filter to the processing pipeline.
+
+        Parameters
+        ----------
+        filter_spec : str or tuple
+            Name of the filter function to add, or a tuple where the first element is the name of the 
+            filter function and the second element is a dictionary containing additional arguments for the function.
+        """
+        if isinstance(filter_spec, str):
+            filter_func = FILTER_FUNCTIONS[filter_spec]
+        elif isinstance(filter_spec, tuple):
+            filter_name, filter_args = filter_spec
+            filter_func = partial(FILTER_FUNCTIONS[filter_name], **filter_args)
+            filter_func.__name__ = FILTER_FUNCTIONS[filter_name].__name__
+        else:
+            raise TypeError("filter_spec should be a string or a tuple")
+
+        self.filters.append(filter_func)
+        # Sort filters according to their order in self.filter_order
+        self.filters.sort(key=lambda f: self.filter_order.index(f.__name__))
+
+    def add_custom_filter(self, filter_function, filter_params=None):
+        """
+        Add a custom filter function to the processing pipeline.
+
+        Parameters
+        ----------
+        filter_function: callable
+            Custom function to add to the processing pipeline.
+            Expects a function that takes a matchms Spectrum object as input and returns a Spectrum object
+            (or None).
+            Regarding the order of execution: the added filter will be executed where it is introduced to the
+            processing pipeline.
+        filter_params: dict
+            If needed, add dictionary with all filter parameters. Default is set to None.
+        """
+        if not callable(filter_function):
+            raise TypeError("Expected callable filter function.")
+        filter_position = 0
+        for filter_func in self.filters[::-1]:
+            if filter_func.__name__ in self.filter_order:
+                filter_position = self.filter_order.index(filter_func.__name__)
+        self.filter_order.insert(filter_position + 1, filter_function.__name__)
+        if filter_params is None:
+            self.filters.append(filter_function)
+        else:
+            filter_func = partial(filter_function, **filter_params)
+            filter_func.__name__ = filter_function.__name__
+            self.filters.append(filter_func)
+
+    def process_spectrum(self, spectrum):
+        """
+        Process the given spectrum with all filters in the processing pipeline.
+
+        Parameters
+        ----------
+        spectrum : Spectrum
+            The spectrum to process.
+        
+        Returns
+        -------
+        Spectrum
+            The processed spectrum.
+        """
+        if self.filters == []:
+            raise TypeError("No filters to process")
+        for filter_func in self.filters:
+            spectrum = filter_func(spectrum)
+            if spectrum is None:
```

Comment:
I would like to keep track of the number of changes each filter step did, instead of what fields were changed, to get a bit more insight into what is happening. This can be implemented by calling processing_report.add_to_report() here. 

Add_to_report does not have to find exactly what was changed in this case. I would instead just return, was changed, or was not changed. Alternatively, we could extend this to spectrum was removed, a field was added, or was changed, but this is often already clear from the function name. 

This can be implemented without changing the filtering functions to classes, but it might be causing some extra overhead since all fields have to be checked again after each filter function. A class approach would make this faster and cleaner, but I agree that this is probably not worth the time investment by us, so this is a nice in-between solution.